crowbar_framework/app/models/api/crowbar.rb

```diff
@@ -80,6 +80,60 @@ def maintenance_updates_installed?
       end
     end
 
+    def compute_resources_available?
+      compute_nodes = NodeObject.find("roles:nova-compute-kvm")
+      return true if compute_nodes.empty?
+      if compute_nodes.size == 1
+        Rails.logger.warn("Only one compute node found; non-disruptive upgrade is not possible!")
+        return false
+      end
+
+      # Now, find out if we have enough memory for migrating
+
+      # Find the controller responsible for the nova services
+      founders = NodeObject.find(
+        "roles:nova-controller AND pacemaker_founder:true AND pacemaker_config_environment:*"
+      )
+      controller = founders.first
+
+      free_compute_memory = 0
+      largest_compute_node_total = 0
+      largest_compute_node_free = 0
+
+      compute_nodes.each do |compute_node|
+        # FIXME: are we sure nova uses the short name of node for its hosts?
+        name = compute_node.handle
+
+        total_cmd = ". /root/.openrc; /usr/bin/nova host-describe #{name} | grep '(total)'"
+        ssh_retval = controller.run_ssh_cmd(total_cmd)
+        out = ssh_retval[:stdout] || ""
+        total_memory = out.split[7].to_i
+
+        used_cmd = ". /root/.openrc; /usr/bin/nova host-describe #{name} | grep '(used_now)'"
+        ssh_retval = controller.run_ssh_cmd(used_cmd)
+        out = ssh_retval[:stdout] || ""
+        used_memory = out.split[7].to_i
+
+        free_memory = total_memory - used_memory
+        free_compute_memory += free_memory
+
+        # find the node with largest total memory
+        if total_memory > largest_compute_node_memory
+          largest_compute_node_total = total_memory
+          largest_compute_node_free = free_memory
+        end
+      end
+
+      if free_compute_memory - largest_compute_node_free < largest_compute_node_total
```

Comment:
Could you explain what you are trying to calculate? You are testing whether the `total free memory` minus the `free memory on the node with the largest total memory` is larger than the `total memory on the node with the largest total memory`. Why? Are you testing whether you can migrate all instances to the largest node during upgrade? In that case I would have expected to test whether the `total used memory` minus the `used memory on the node with the largest free memory` is less than the `free memory on the node with the largest free memory`.